crypto/ffc/ffc_params_validate.c

```diff
@@ -111,7 +115,39 @@ int ossl_ffc_params_simple_validate(OSSL_LIB_CTX *libctx, FFC_PARAMS *params,
 #endif
         ret = ossl_ffc_params_FIPS186_4_validate(libctx, params, type, &res,
                                                  NULL);
+#ifndef OPENSSL_NO_DH
+    if (ret == FFC_PARAM_RET_STATUS_FAILED
+        && (res & FFC_ERROR_NOT_SUITABLE_GENERATOR) != 0) {
+        ERR_raise(ERR_LIB_DH, DH_R_NOT_SUITABLE_GENERATOR);
+    }
+#endif
     params->flags = save_flags;
     params->gindex = save_gindex;
+
+    /*
+     * This is called also when generating ephemeral DH keys.
+     * Let's avoid an expensive primality checks in that case.
+     */
+    if (ret != FFC_PARAM_RET_STATUS_FAILED && type == FFC_PARAM_TYPE_DSA) {
+        BN_CTX *ctx;
+
+        if ((ctx = BN_CTX_new_ex(libctx)) == NULL)
+            return 0;
+        if (BN_check_prime(params->p, ctx, NULL) != 1) {
```

Comment:
I wondered if these were too slow to put here..

Reply:
It is used as a partial check of the params in both dsa and dh keygen.

Reply:
It is commented about up there. The prime checks will be called only when generating DSA keys which are not ephemeral so I do not see any issue with the slowdown.

There is one related question though - is there any way how to generate DSA parameters that will pass the g validation check but not the primality test for the p value? @romen could perhaps know the answer? Or in other words, is the primality check for p really worth it?

Reply:
Keep in mind this PR also..
https://github.com/openssl/openssl/pull/14146

Reply:
The question is how quick the quick check is supposed to be. Whether the g validation (with the modular exponentiation call) is quick enough for the quick check.